File: python/tvm/relax/op/ccl/ccl.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Relax Collective Communications Library (CCL) operators"""
from typing import Union
from tvm.relax import PrimValue

from . import _ffi_api
from ...expr import Expr
from ....ir import PrimExpr


def allreduce(x, op_type: str = "sum"):  # pylint: disable=invalid-name
    """Allreduce operator

    Parameters
    ----------
    x : relax.Expr
      The input tensor.
    op_type: str
      The type of reduction operation to be applied to the input data.
      Now "sum", "prod", "min", "max" and "avg" are supported.

    Returns
    -------
    result : relax.Expr
      The result of allreduce.
    """
    supported_op_types = ["sum", "prod", "min", "max", "avg"]
    assert op_type in supported_op_types, (
        "Allreduce only supports limited reduction operations, "
        f"including {supported_op_types}, but got {op_type}."
    )
    return _ffi_api.allreduce(x, op_type)  # type: ignore # pylint: disable=no-member


def allgather(x, num_workers: Union[int, PrimExpr, PrimValue]):  # pylint: disable=invalid-name
    """AllGather operator

    Parameters
    ----------
    x : relax.Expr
      The input tensor.

    num_worker : Union[int, PrimExpr, PrimValue]
      The number of workers to gather data from.

    Returns
    -------
    result : relax.Expr
      The result of allgather.
    """
    if not isinstance(num_workers, PrimValue):
        num_workers = PrimValue(num_workers)
    return _ffi_api.allgather(x, num_workers)  # type: ignore # pylint: disable=no-member


def broadcast_from_worker0(x: Expr) -> Expr:
    """Broadcast data from worker-0 to all other workers.

    Parameters
    ----------
    x : relax.Expr
      The tensor to be broadcast.

    Returns
    -------
    result : relax.Expr
      The same tensor, which has been broadcast to all other workers.
    """
    return _ffi_api.broadcast_from_worker0(x)


def scatter_from_worker0(x: Expr, num_workers: int, axis: int = 0) -> Expr:
    """Perform a scatter operation from worker-0, chunking the given buffer into equal parts.

    Parameters
    ----------
    x : relax.Expr
      The buffer to be divided into equal parts and sent to each worker accordingly.

    num_worker : int
      The number of workers, i.e. the number of parts the given buffer should be chunked into.

    axis : int
      The dimension of the tensor to be scattered. Default is 0.

    Returns
    -------
    result : relax.Expr
      Chunked Tensor received by different workers.
    """
    return _ffi_api.scatter_from_worker0(x, num_workers, axis)

File: src/relax/op/ccl/ccl.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "ccl.h"

#include <tvm/ffi/extra/visit_error_context.h>
#include <tvm/ffi/reflection/registry.h>
#include <tvm/relax/block_builder.h>

#include <utility>

namespace tvm {
namespace relax {

/* relax.ccl.allreduce */

TVM_FFI_STATIC_INIT_BLOCK() {
  AllReduceAttrs::RegisterReflection();
  AllGatherAttrs::RegisterReflection();
  ScatterCollectiveAttrs::RegisterReflection();
}

Expr allreduce(Expr x, ffi::String op_type, bool in_group) {
  ffi::ObjectPtr<AllReduceAttrs> attrs = ffi::make_object<AllReduceAttrs>();
  attrs->op_type = std::move(op_type);
  attrs->in_group = std::move(in_group);

  static const Op& op = Op::Get("relax.ccl.allreduce");
  return Call(op, {std::move(x)}, Attrs{attrs}, {});
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("relax.op.ccl.allreduce", allreduce);
}

Type InferTypeAllReduce(const Call& call, const BlockBuilder& ctx) {
  TensorType input_ty = GetUnaryInputTensorType(call, ctx);
  return input_ty;
}

TVM_REGISTER_OP("relax.ccl.allreduce")
    .set_attrs_type<AllReduceAttrs>()
    .set_num_inputs(1)
    .add_argument("x", "Tensor", "Input to which allreduce will be applied.")
    .set_attr<FInferType>("FInferType", InferTypeAllReduce)
    .set_attr<FRelaxInferLayout>("FRelaxInferLayout", InferLayoutUnaryEwise)
    .set_attr<bool>("FPurity", true);

/* relax.ccl.allgather */

Expr allgather(Expr x, int num_workers, bool in_group) {
  ffi::ObjectPtr<AllGatherAttrs> attrs = ffi::make_object<AllGatherAttrs>();
  attrs->num_workers = std::move(num_workers);
  attrs->in_group = std::move(in_group);

  static const Op& op = Op::Get("relax.ccl.allgather");
  return Call(op, {std::move(x)}, Attrs{attrs}, {});
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("relax.op.ccl.allgather", allgather);
}

Type InferTypeAllGather(const Call& call, const BlockBuilder& ctx) {
  TensorType input_ty = GetUnaryInputTensorType(call, ctx);

  const auto* attrs = call->attrs.as<AllGatherAttrs>();
  int num_workers = attrs->num_workers;

  PrimType output_dtype = input_ty->dtype;
  auto input_shape = input_ty->GetShape();
  if (!input_shape.defined()) {
    return input_ty;
  }
  ffi::Array<PrimExpr> output_shape = input_shape.value();
  output_shape.Set(0, floor(output_shape[0] * num_workers));
  return TensorType(ShapeExpr(output_shape), output_dtype, input_ty->vdevice);
}

TVM_REGISTER_OP("relax.ccl.allgather")
    .set_num_inputs(1)
    .add_argument("x", "Tensor", "Input to which allgather will be applied.")
    .set_attr<FInferType>("FInferType", InferTypeAllGather)
    .set_attr<FRelaxInferLayout>("FRelaxInferLayout", InferLayoutUnaryEwise)
    .set_attr<bool>("FPurity", true);

/* relax.ccl.broadcast_from_worker0 */
Expr broadcast_from_worker0(Expr x) {
  static const Op& op = Op::Get("relax.ccl.broadcast_from_worker0");
  return Call(op, {std::move(x)}, {}, {});
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("relax.op.ccl.broadcast_from_worker0", broadcast_from_worker0);
}

Type InferTypeBroadcastFromZero(const Call& call, const BlockBuilder& ctx) {
  TensorType input_ty = GetUnaryInputTensorType(call, ctx);
  return input_ty;
}

TVM_REGISTER_OP("relax.ccl.broadcast_from_worker0")
    .set_num_inputs(1)
    .add_argument("x", "Tensor", "Input to be broadcast.")
    .set_attr<FInferType>("FInferType", InferTypeBroadcastFromZero)
    .set_attr<FRelaxInferLayout>("FRelaxInferLayout", InferLayoutUnaryEwise)
    .set_attr<bool>("FPurity", true);

/* relax.ccl.scatter_from_worker0 */

Expr scatter_from_worker0(Expr data, int num_workers, int axis) {
  ffi::ObjectPtr<ScatterCollectiveAttrs> attrs = ffi::make_object<ScatterCollectiveAttrs>();
  attrs->num_workers = std::move(num_workers);
  attrs->axis = std::move(axis);
  static const Op& op = Op::Get("relax.ccl.scatter_from_worker0");

  return Call(op, {std::move(data)}, Attrs{attrs}, {});
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef().def("relax.op.ccl.scatter_from_worker0", scatter_from_worker0);
}

Type InferTypeScatter(const Call& call, const BlockBuilder& ctx) {
  TensorType input_ty = GetUnaryInputTensorType(call, ctx);
  PrimType output_dtype = input_ty->dtype;

  const auto* attrs = call->attrs.as<ScatterCollectiveAttrs>();
  int num_workers = attrs->num_workers;

  arith::Analyzer analyzer = ctx->GetAnalyzer();
  auto input_shape = input_ty->GetShape();
  TVM_FFI_ICHECK(input_shape.defined())
      << "input tensor of scatter_from_worker0 should have defined shape.";

  if (analyzer->CanProve(floormod(input_shape.value()[attrs->axis], PrimExpr(num_workers)) != 0)) {
    TVM_FFI_VISIT_THROW(ValueError, call)
        << "scatter_from_worker0 expects the size of axis " << attrs->axis
        << " of input tensor to be divisible by the num_workers. However, axis " << attrs->axis
        << " of input tensor is " << input_shape.value() << " while num_workers is " << num_workers;
  }

  ffi::Array<PrimExpr> output_shape = input_shape.value();
  output_shape.Set(attrs->axis, div(output_shape[attrs->axis], num_workers));
  return TensorType(ShapeExpr(output_shape), output_dtype, input_ty->vdevice);
}

TVM_REGISTER_OP("relax.ccl.scatter_from_worker0")
    .set_num_inputs(1)
    .add_argument("x", "Tensor",
                  "The buffer to be divided into equal parts and sent to each worker accordingly.")
    .set_attrs_type<ScatterCollectiveAttrs>()
    .set_attr<FInferType>("FInferType", InferTypeScatter)
    .set_attr<bool>("FPurity", true);

}  // namespace relax
}  // namespace tvm
